app/editor/src/features/content/list-view/utils/naturalSort.ts

```diff
@@ -0,0 +1,19 @@
+import { IContentSearchResult } from 'store/slices';
+
+function getPageSectionValue(row: IContentSearchResult) {
+  // gnerate page:section string, keep it in the lowercase
+  const value = `${row.original.page ? row.original.page : ''}:${
+    row.original.section ? row.original.section : ''
+  }`.toLowerCase();
+  return value;
+}
+
+export function naturalSortValue(row: IContentSearchResult) {
+  const pageSectionValue = getPageSectionValue(row);
+  // Replace each segment of digits and non-digits with formatted strings
+  // Digits are padded with zeros to the left to ensure correct natural sorting
+  // Non-digits are left as is
+  return pageSectionValue.replace(/(\d+)|(\D+)/g, (_, $1, $2) =>
```

Comment:
If I'm reading this correct you're extracting the decimal value from the string, and then replacing it by padding the number with 10 zeros?

What I don't understand is how this will be applied to the list-view?  I'm not seeing the code that links this.  How would this be tested currently?

Reply:
Hi Jeremy. 
It added zero to number to help sort.

```
> testData = {
...   original: {
...     page: 'Page2',
...     section: 'sport'
...   }
... };
{ original: { page: 'Page2', section: 'sport' } }
> testData2 = {
...   original: {
...     page: 'Page02',
...     section: 'sport'
...   }
... };
{ original: { page: 'Page02', section: 'sport' } }
>
> naturalSortValue(testData)
'page0000000002:sport'
> naturalSortValue(testData2)
'page0000000002:sport'
```

so the "page2" and "page02" can convert to a same string.


The problem I'm currently facing is:

**A.** Normally, I cannot pass a function into the sort function. This cannot be executed, and `col.sort` will be `undefined`. to use sort as a function this involves modifying the code in tno-core.
First, I modified the code in [useColumns.tsx](https://github.com/bcgov/tno/blob/dev/app/editor/src/features/content/list-view/hooks/useColumns.tsx#L70) the editor list view:

Passing a function to sort:

  `accessor: 'section',   sort: (row) => naturalSortValue(row)`

In [flexboxTable.tsx,](https://github.com/bcgov/tno/blob/dev/libs/npm/core/src/components/table/FlexboxTable.tsx#L112-L115) 
```sort:
  col.sort ?? typeof col.accessor === 'function'
    ? undefined
    : col.accessor,
```

I replace it by the below
```
const determineSortValue = (col: ITableInternalHeaderColumn<T>) => {
    // If 'col.sort' is defined and not null, use it directly for sorting
    if (col.sort !== undefined && col.sort !== null) {
      console.log('col.sort', col.sort);
      return col.sort;
    }
    // If 'col.sort' is undefined or null and 'accessor' is a function,
    // it indicates that 'accessor' cannot be used for sorting directly (e.g., dynamic values)
    if (typeof col.accessor === 'function') {
      console.log('col.accessor', col.accessor);
      return undefined;
    }
    // If 'col.sort' is undefined or null, and 'accessor' is not a function,
    // use 'accessor' as the field name for sorting
    return col.accessor;
}

sort: determineSortValue(col)

```



This let `sort` can be a function, and also allow us to use sort function in  [sortRows.ts](https://github.com/bcgov/tno/blob/dev/libs/npm/core/src/components/table/utils/sortRows.ts#L36-L41)

to execute the following code block :

```

      if (sortBy) {
        if (typeof sortBy.sort === 'function') {
          const aVal = sortBy.sort(a) ?? '';
          const bVal = sortBy.sort(b) ?? '';
          return (aVal < bVal ? -1 : 1) * (sortBy.isSortedDesc ? -1 : 1);
        }



```
But I'm not sure if we want to make these changes. it may affect other parts of the code.


**B.** If I pass a function into the accessor, it will cause an Elasticsearch error later on.

such as simply replace accessor [useColumns.tsx](https://github.com/bcgov/tno/blob/dev/app/editor/src/features/content/list-view/hooks/useColumns.tsx#L69) the editor list view:

`accessor: 'section'`  ===> `accessor : (row) => naturalSortValue(row)`

 in fact, using any function here will always cause an error

Reply:
Elasticsearch will not understand any of this fancy sorting.  The solution that will work for us is to simply sort the results in the UI after Elasticsearch returns the results.

Reply:
It would be a good idea to apply to new content coming in, but isn't critical for this fix.  Applying the sort to the UI should resolve the issue 99% of the time.

Reply:
In fact, you can test it. I think it's fine in most cases. Now it can sort resource first then page and section